exonum/src/storage/proof_list_index/mod.rs

```diff
@@ -614,3 +618,33 @@ where
         self.base_iter.next().map(|(_, v)| v)
     }
 }
+
+/// Computes Merkle root hash for a given list of hashes.
+///
+/// If `hashes` are empty then `Hash::zero()` value is returned.
+pub fn root_hash(hashes: &[Hash]) -> Hash {
+    match hashes.len() {
+        0 => Hash::zero(),
+        1 => hashes[0],
+        _ => {
+            let mut current_hashes = combine_hash_list(hashes);
+            while current_hashes.len() > 1 {
+                current_hashes = combine_hash_list(&current_hashes);
+            }
+            current_hashes[0]
+        }
+    }
+}
+
+fn combine_hash_list(hashes: &[Hash]) -> Vec<Hash> {
+    hashes
+        .chunks(2)
+        .map(|pair| {
```

Comment:
Nit: you can use slice comprehension here:

```rust
match pair {
    [first, second] => hash_pair(first, second),
    [single] => hash_one(single),
    _ => unreachable!(),
}
```

Reply:
I also think that this looks clearer than `if/else`.